### ISSUE TITLE
Implement conditional operators

### ISSUE THREAD

**Issue by rebolbot** [rebol/rebol-issues#1879]
_Submitted by:_ **Ladislav**

By "conditional expressions" I mean expressions that are used to supply CONDITION arguments to control functions like IF, EITHER, UNLESS, ...

Conditional expressions can also be found in BLOCK arguments of control functions like ANY, ALL, CASE, WHILE, UNTIL.

In programming languages the purpose of operators like NOT, AND, OR, XOR is to combine simple conditional expressions yielding complex conditional expressions.

However, in Rebol the only conditional operator I know is NOT, which is able to take a result of a conditional expression as an argument and yield a "conditional result".

That is not the case of the AND, OR and XOR operators, which cannot take conditional values as arguments. (see the example expression below)

The fact that NOT is conditional, while AND, OR and XOR are not makes the operators mutually incompatible. That confuses users not aware of such catches. My findings prove that too many users are being caught by that.

BrianH suggested that nonconditional AND, OR and XOR operators are better than conditional, since:

* they can be used in conditional expressions with a help of the TRUE? function

* they protect the user, since they "cut down on none propagation"

There are serious objections to both of these, though:

* users don't remember the differences between TRUE?, TO-LOGIC and FOUND?, using the most inappropriate of the three functions too often

* the expressivity of the language suffers, since the users don't want to replace conditional expressions like

``` rebol
    if value1 AND value2 OR value3 []
```

by

``` rebol
    if (true? value1) AND (true? value2) OR (true? value3) []
```

* the "cut down on none propagation" property actually does not work for conditional expressions like the latter one anyway

* virtually nobody knows that the counterpart of nonconditional AND, OR and XOR operators is not the conditional NOT operator, but the nonconditional COMPLEMENT operator. The usage of the COMPLEMENT operator is known to be much lower than the usage of the NOT operator. Notice however that if the users wanted to enjoy the purported benefits of the nonconditional operators, they would use the COMPLEMENT operator more frequently.

== Summary

* the conditional NOT operator is preferred by the users, when compared to its nonconditional COMPLEMENT variant (because, what the users want is to have a conditional operator, not a nonconditional one)

* Rebol seems to be currently the only programming language not having conditional AND and OR operators

== Possible solution

* Make the AND, OR and XOR operators conditional

* Similarly as for the NOT operator, the AND, OR and XOR operators already have nonconditional counterparts. Naturally, the nonconditional AND counterpart is INTERSECT (currently just duplicating some of the functionality), the natural nonconditional counterpart of OR is UNION, and the natural nonconditional counterpart of XOR is DIFFERENCE.

* My user poll results suggest that the majority of users would welcome this change.

    ; OR cannot combine conditional subexpressions,
    ; the expression below triggers a script error
    none or none

<sup>**CC - Data** [ Version: alpha 111 Type: Wish Platform: All Category: Native Reproduce: Always Fixed-in:none ]</sup>

**Comment by rebolbot** [comment]
_Submitted by:_ **BrianH**

> "Is there any other conditional operator in Rebol?"

You are proposing in this ticket that AND, OR and XOR be changed to conditional. So, for the existing Rebol behavior before this change would be implemented, substitute "logical" for "conditional" in that comment (actually, I'll edit the comment to clarify that is what I meant).

And yes, the "not" operator has a conditional (or logical if you prefer) variant (NOT) and a bitwise variant (COMPLEMENT). That is similar to what would have to be done for AND, OR and XOR before we could go forward with this proposal. Make a separate proposal for those new bitwise operators and we will be able to move forward with both.

**Comment by rebolbot** [comment]
_Submitted by:_ **Ladislav**

Having refreshed my memory, the current behaviour is as follows:

* AND is a logical and bitwise operator

the change proposed above is to:

* make AND a conditional operator
* make INTERSECT a bitwise operator

similarly for the OR (UNION) and XOR (DIFFERENCE) operators

**Comment by rebolbot** [comment]
_Submitted by:_ **Ladislav**

To discuss it further, assuming that a consensus to change AND, OR and XOR to conditional operators is more or less present, I assume that:

* AND~, OR~ and XOR~ prefix versions should also be transformed to conditional operators
* INTERSECT, UNION and DIFFERENCE should be transformed to work also as logic operators and bitwise operators for binaries and integers

**Comment by rebolbot** [comment]
_Submitted by:_ **abolka**

Noting that I haven't expressly registered my agreement here: +1

* +1 for making infix AND, OR, XOR conditional
* +1 for making prefix AND~, OR~, XOR~ conditional as well
* +1 for introducing a separate set of infix bitwise operators for binaries and integers (AND*, OR+, XOR-; or AND#, OR#, XOR#)

**Comment by hostilefork** [comment]
**This is mostly implemented in Ren-C, with outstanding issues below.**

> Similarly as for the NOT operator, the AND, OR and XOR operators already have nonconditional counterparts. Naturally, the nonconditional AND counterpart is INTERSECT (currently just duplicating some of the functionality), the natural nonconditional counterpart of OR is UNION, and the natural nonconditional counterpart of XOR is DIFFERENCE.

Sounds good on digital paper...and I was going with this in Ren-C.  Until I found it's not *quite* true, e.g. there are different meanings to the set operations and bitwise operations in an input they accept in common: BINARY!

    >> intersect #{FFFF} #{FF00}
    == #{FF}

    >> and~ #{FFFF} #{FF00}      
    == #{FF00}

It depends on whether you consider a BINARY! as a "set of bytes" (the way strings are treated as sets of codepoints) or if you consider it a series of bits to be processed with bitwise boolean logic.

That problem arose while implementing the given proposal in Ren-C, and I do not as yet have a solution for it.  But other things are fairly well set.

Choice for new namings of the bitwise infix operators is `AND*`, `OR+`, and `XOR-` ... which hints at their "mathy" function and also is consistent with the definitions.  e.g. bitwise AND is considered boolean's "multiply" because 1 is the identity `(1 AND X AND Y) = (X AND Y)` as `(1 * X * Y) = (X * Y)` and 0 is the identity in OR as boolean's addition `(0 OR X OR Y) = (X OR Y)` as `(0 + X + Y) = (X + Y)`.  *(Relating XOR to subtraction has some precedent too, I think...though more abstractly...and I'm not sure what one might put on the end of a NOT to make it mathy or if there just won't be one and you call COMPLEMENT)*

As for prefix conditional, `AND?`, `OR?` and `XOR?` as proposed above are also in Ren-C.  A `NOT?` might be added for good measure though it would be a synonym for NOT.  Leading to holding off on that decision was the question of if `AND [X Y]` would be treated as `ALL [X Y]` and fit into the NONE vs. conditionally-true outcome, or if it would return TRUE or FALSE.  But prudence, predictability, and realism dictate that e.g. `true and false` be FALSE, so it is better to force to LOGIC! even if that is "lossy".

**Comment by hostilefork** [comment]
This has been implemented in Ren-C, with the acknowledgment that it means that INTERSECT/etc. of BINARY! will behave differently.  Considered worth the tradeoff to get rid of the AND~, OR~, XOR~.

If someone clamors for a refinement to intersect binaries as sets of bytes, it is trivial to allow it.

**Comment by hostilefork** [comment]
**Note: While this contains the gist of the change, several improvements have been made since.  Please see [the Trello card](https://trello.com/c/oweGGIgf/222-and-and-or-are-conditional-and-can-be-used-for-logic-driven-branching).  Users can choose to put the right hand side in a block or a group, and if they use a group then they get non-short-circuit evaluation...but with the same value result of the overall operation.**

Taking this popular user suggestion and implementing it--brought something to light over time that hadn't been discussed yet.  @Ladislav said:

> "Rebol seems to be currently the only programming language not having conditional AND and OR operators"

While that is true, we should notice as well that "most" languages have *short-circuit* AND and OR operators.  Consider something like:

     b: 1020
     either (block? b) and (304 = length? b) [
         print "b is a block of length 304"
     ][
         print "b is either not a block, or does not have length 304"
     ]

If the motivation of the change is to give users something that feels more natural...I'd imagine the "natural" expectation would be to get the second message.  But with this proposal unmodified, it would error...since it's not short circuit.  So both sides of the AND would be evaluated, and you can't run `length? 1024`.

This points to one advantage that defining boolean operators as bitwise vs. conditional has.  It avoids people getting bitten by their non-short-circuiting nature.

Yet we don't have to declare the issue a death-sentence for this popular-feeling proposal.  While languages like C# need things like "lambda lifting" to get short circuit operator overloads, when other parts of Rebol face this problem...such as the branches of conditionals, they just put the code into a BLOCK!:

     b: 1020
     either (block? b) and [304 = length? b] [
         print "b is a block of length 304"
     ][
         print "b is either not a block, or does not have length 304"
     ]

The lack of symmetry is a bit off-putting, however.  It can be explained, but it "looks funny".  Nearly every expression on the left would need to be parenthesized in R3-Alpha due to the rules of left-hand infix.  And every expression on the right would need to be in a block.

But Rebol's point is being bendable enough to dress up in the illusion you would like.  If that weren't the point, there'd be no reason for infix at all...it would all be prefix.  The "user-friendly" infix form of AND doesn't have to follow any particular interface rules...there's already AND? if you want non-short circuit behavior, and ALL if you need something more generic.

Using Ren-C's more generalized "left-enfix normality", we can have a short-circuiting operator that merely asks you to put your right hand side in a GROUP!:

    and: enfix func [
        left [any-value!]
            {Expression which will always be evaluated}
        :right [group!]
            {Quoted expression, that will be evaluated only if LEFT is TRUTHY?}
    ][
        either left [to-logic do right] [false]
    ]

This allows you to write fairly normal looking expressions, so long as the right side is in a GROUP!, and get short-circuiting:

     b: 1020
     if block? b and (length of b = 304) and (...) [
         print "b is a block of length 304"
     ] else [
         print "b is either not a block, or does not have length 304"
     ]

Due to the workings of left normal enfix, the GROUP! also means it can chain.  Function calls on the left must be completed before the right can run, so left conditions will short circuit arbitrarily long chains on the right.

This gives the user the requirement to put the right hand side in a GROUP!, but the option to put the left hand side in a GROUP!.  Expressively it can look very natural, blending with some of the other illusions...and having a fairly easy-to-explain logic behind it.  Not requiring a GROUP! on the left fits with the principle of minimum boilerplate...which drives the idea of why the condition of an IF doesn't need to be in a BLOCK!, while the condition of a WHILE does.

Just to cover all the ground: it's worth pointing out that not all languages--especially extensible ones--have means of doing short-circuiting in cases like these.  C++ for example, [does not short-circuit overloaded `&&` or `||`](https://stackoverflow.com/questions/25913237/is-there-actually-a-reason-why-overloaded-and-dont-short-circuit).  And since as a language Rebol already has ANY and ALL for doing short circuiting, there could be a distinct value to having operators that do *not* short circuit.

So it would simply be a case of thinking about whether you want both sides of the expression to be evaluated or not.  Certainly that's a useful distinction for code golf.  But this may feel like a level of awareness that only a "power user" is going to be able to fathom...which undermines the argument of the change aiming at "user friendliness" in reading/writing.

Point is, there are options available here.

1. Abandon the idea of conditional AND/OR and go back to bitwise math.  I feel that's lame.
2. Keep the idea but live with the non-short-circuit behavior...always evaluating both sides.  I could get used to it, and it would have its applications, but it's not exactly a slam dunk over leaving the bitwise behavior...part of the argument for this was that it would *help* new users, not confuse them more.
3. Do it short circuit and have right-hand-side expressions in **quoted** GROUP!s...it looks good enough, while being a little sneaky.  But sneakiness to achieve an illusion is part of what makes Rebol interesting.  And people might have wanted to put the expression in a group in the first place anyway.
4. Do it short circuit and have right-hand-side expressions in BLOCK!s.  I think this makes the code look a little "chunky" and less "balanced" than I'd like...when the operator is there for visual/layout reasons.  Also it would be strange to hard-quote a BLOCK!...you'd *have* to hard quote it, because allowing **foo: [...] bar: [...] if foo and bar [...]** would treat the first foo conditionally, and run the second as code.   So GROUP!s might not just be better looking, but also be less questioned over the hard quote, since it's obviously required for deferred execution.

Ren-C is [going with option #3](https://github.com/hostilefork/rebol/commit/91b92ddae5f4b097992831c5ab419f1a92501acf).

---

**UPDATE:** The actual mechanics wound up distinguishing quoted GROUP! on the right from quoted BLOCK! on the right, in a finesse that is really rather remarkably cool.  GROUP! on the right implies forcing the result to logic, while BLOCK! implies value-preservation...more like an ANY or ALL would do.  This makes OR [...] a very nice tool as a logic-driven analogue to ELSE.  See:

https://trello.com/c/oweGGIgf

**Comment by hostilefork** [comment]
This is *largely* resolved in Ren-C, with two outstanding issues:

[Cannot treat BINARY! as set of bytes via AND+, OR+, XOR+](https://github.com/metaeducation/ren-c/issues/700)

[Getting time delta on DATE!s should not use DIFFERENCE](https://github.com/rebol/rebol-issues/issues/2287)

Note also this test from Rebol2 does not attempt to bitwise intersect individual characters in a string, but intersects the strings as sets:

      "^(03)^(00)" and+ "^(02)^(00)" = "^(00)" ;-- Rebol2 expected "^(02)^(00)"
